cmds/dhclient/dhclient.go

```diff
@@ -30,33 +32,48 @@ import (
 )
 
 const (
-	defaultIfname = "eth0"
 	// slop is the slop in our lease time.
-	slop = 10 * time.Second
+	slop          = 10 * time.Second
+	linkUpAttempt = 30
 )
 
 var (
+	ifName       = "^e.*"
 	leasetimeout = flag.Int("timeout", 15, "Lease timeout in seconds")
 	retry        = flag.Int("retry", -1, "Max number of attempts for DHCP clients to send requests. -1 means infinity")
-	renewals     = flag.Int("renewals", -1, "Number of DHCP renewals before exiting")
-	verbose      = flag.Bool("verbose", true, "Verbose output")
+	renewals     = flag.Int("renewals", -1, "Number of DHCP renewals before exiting. -1 means infinity")
+	verbose      = flag.Bool("verbose", false, "Verbose output")
 	ipv4         = flag.Bool("ipv4", false, "use IPV4")
 	test         = flag.Bool("test", false, "Test mode")
 	debug        = func(string, ...interface{}) {}
 )
 
 func ifup(ifname string) (netlink.Link, error) {
-	iface, err := netlink.LinkByName(ifname)
-	if err != nil {
-		return nil, fmt.Errorf("cannot get interface by name %v: %v", ifname, err)
-	}
-
-	if err := netlink.LinkSetUp(iface); err != nil {
-		return nil, fmt.Errorf("%v: %v can't make it up: %v", ifname, iface, err)
-	}
+	debug("Try bringing up %v", ifname)
+	start := time.Now()
+	for int(time.Since(start).Seconds()) < linkUpAttempt {
+		// Note that it may seem odd to keep trying the
+		// LinkByName operation, by consider that a hotplug
+		// device such as USB ethernet can just vanish.
+		iface, err := netlink.LinkByName(ifname)
+		debug("LinkByName(%v) returns (%v, %v)", ifname, iface, err)
+		if err != nil {
+			debug("%v not found: %v", ifname, err)
+			return nil, fmt.Errorf("cannot get interface by name %v: %v", ifname, err)
+		}
 
-	return iface, nil
+		if iface.Attrs().OperState == netlink.OperUp {
+			debug("Link %v is up", ifname)
+			return iface, nil
+		}
 
+		if err := netlink.LinkSetUp(iface); err != nil {
+			debug("Could not bring %v up: %v", ifname, err)
```

Comment:
yeah. I was in the middle of some desperate debugging and needed more info that I was getting but I'll remove some of these.